[MLv2] Lay foundations and port display locking logic to CLJC

bshepherdson opened the pull request:
This lays the groundwork for how a `Question` TS class can hold an
immutable CLJS map and immutable JS object holding the card data, and
convert lazily between them as needed.

This also fixes `mu/defn` for CLJS - the original implementation tries
to put a `defn` in a `let` binding and then return it. Vars aren't real
in CLJS, and while this doesn't cause an error during shadow-cljs
compilation, the resulting JS code is slightly broken.

metamben commented on src/metabase/lib/query.cljc:
```diff
@@ -91,3 +92,10 @@
   [query :- ::lib.schema/query]
   (:lib/metadata query))
+
+(mu/defn source-table-id :- ::lib.schema.id/table
```
Cannot this return `nil`?

bshepherdson replied:
This is removed now.

metamben reviewed (approved):
Were the databases really meant to be committed?

bshepherdson committed:
[MLv2] Round-trip JS <-> CLJS Cards on demand; use for display locking

This introduces a Malli schema for `Card`, the representation for a
"question" in Metabase, and supports conversion between the JS map
`Card` and CLJS maps.

The `dataset_query` field holding the native or MBQL query is special,
and this PR uses the `->legacy-MBQL` conversion from #29419.

A TS `Question` can be created with either the JS or CLJS rep for the
card, and the methods `Question.card()` and `Question._cljsCard()`
lazily convert the other form into the requested form and cache it.

The existing `_getMLv2Query` logic is replaced; it is subsumed by the
conversion of the whole card including the query.
